packages/jupyter-ai/src/components/settings/existing-api-keys.tsx

```diff
@@ -0,0 +1,234 @@
+
+export type ExistingApiKeysProps = {
+  alert: StackingAlert;
+  apiKeys: string[];
+  onSuccess: () => unknown;
+};
+
+/**
+ * Component that renders a list of existing API keys. Each API key is rendered
+ * by a unique `ExistingApiKey` component.
+ */
+export function ExistingApiKeys(props: ExistingApiKeysProps): JSX.Element {
+  // current editable API key name, if any.
+  const [editableApiKey, setEditableApiKey] = useState<string | null>(null);
+
+  return (
+    <Box
+      sx={{
+        '& > .MuiBox-root:not(:first-child)': {
+          marginTop: -2
+        }
+      }}
+    >
+      {props.apiKeys.map(apiKey => (
+        <ExistingApiKey
+          key={apiKey}
+          alert={props.alert}
+          apiKey={apiKey}
+          editable={editableApiKey === apiKey}
+          setEditable={setEditableApiKey}
+          onSuccess={props.onSuccess}
+        />
+      ))}
+      {props.alert.jsx}
+    </Box>
+  );
+}
+
+type ExistingApiKeyProps = {
+  alert: StackingAlert;
+  apiKey: string;
+  editable: boolean;
+  setEditable: React.Dispatch<React.SetStateAction<string | null>>;
+  onSuccess: () => unknown;
+};
+
+/**
+ * Component that renders a single existing API key specified by `props.apiKey`.
+ * Includes actions for editing and deleting the API key.
+ */
+function ExistingApiKey(props: ExistingApiKeyProps) {
+  const [input, setInput] = useState('');
+  const [inputVisible, setInputVisible] = useState(false);
+  const [error, setError] = useState(false);
+  const inputRef = useRef<HTMLInputElement>();
+
+  /**
+   * Effect: Select the input after `editable` is set to `true`. This needs to
+   * be done in an effect because the TextField needs to be rendered with
+   * `disabled=false` first. When `editable` is set to `false`, reset any
+   * input-related state.
+   */
+  useEffect(() => {
+    if (props.editable) {
+      inputRef.current?.focus();
+    } else {
+      setInput('');
+      setInputVisible(false);
+      setError(false);
+    }
+  }, [props.editable]);
+
+  const onEditIntent = useCallback(() => {
+    props.setEditable(props.apiKey);
+  }, []);
+
+  const onDelete = useCallback(() => {
+    return AiService.deleteApiKey(props.apiKey);
+  }, []);
+
+  const toggleInputVisibility = useCallback(() => {
+    setInputVisible(visible => !visible);
+  }, []);
+
+  const onEditCancel = useCallback(() => {
+    props.setEditable(null);
+  }, []);
+
+  const onEditSubmit = useCallback(() => {
+    return AiService.updateConfig({
+      api_keys: { [props.apiKey]: input }
+    });
+  }, [input]);
+
+  const onError = useCallback(
+    emsg => {
+      props.alert.show('error', emsg);
+    },
+    [props.alert]
+  );
+
+  const validateInput = useCallback(() => {
+    if (!props.editable) {
+      return;
+    }
+
+    setError(!input);
+  }, [props.editable, input]);
+
+  const onEditSuccess = useCallback(() => {
+    props.setEditable(null);
+    props.alert.show('success', 'API key updated successfully.');
```

Comment:
In a lot of these messages, the word `successfully` seems redundant. Since the left panel is already space-constrained, could these be shortened to, for example, `API key updated.`?

Reply:
I prefer being explicit over being implicit.